rust/src/qfunction.rs

```diff
@@ -404,6 +404,16 @@ impl QFunction {
         unsafe {
             bind_ceed::CeedQFunctionAddInput(self.qf_core.ptr, name_c.as_ptr(), size, emode);
         }
+        /* Note: adding an input can change the pointer to self, resulting in bad ctx data */
+        unsafe {
+            bind_ceed::CeedQFunctionContextSetData(
+                self.qf_ctx_ptr,
+                crate::MemType::Host as bind_ceed::CeedMemType,
+                crate::CopyMode::UsePointer as bind_ceed::CeedCopyMode,
+                10, /* Note: size not relevant - CPU only approach */
+                &mut self as *mut _ as *mut ::std::os::raw::c_void,
+            );
+        }
```

Comment:
This was annoying to find

Reply:
Good catch, but I think this will be a recurring nightmare. We're creating a self-referential data structure so the context should be pinned. https://doc.rust-lang.org/nightly/core/pin/index.html

Reply:
Will this be an issue though
> Instead, Pin<P> prevents certain values (pointed to by pointers wrapped in Pin<P>) from being moved by making it impossible to call methods that require &mut T on them (like mem::swap).

Will revisit tomorrow

Reply:
Ah, you beat me to it. That's what I was thinking of doing too

Reply:
I think this is resolved now.